src/math/WadRayMath.sol

```diff
@@ -116,4 +116,70 @@ library WadRayMath {
             }
         }
     }
+
+    /// @notice Computes the wad-based weighted average (x * (1 - p) + y * p), rounded half up.
+    /// @param x The first value, with a weight of 1 - weight.
+    /// @param y The second value, with a weight of weight.
+    /// @param weight The weight of y, and complement of the weight of x (in wad).
+    /// @return z The result of the wad-based weighted average.
+    function wadWeightedAvg(
+        uint256 x,
+        uint256 y,
+        uint256 weight
+    ) internal pure returns (uint256 z) {
+        // Must revert if
+        //     weight > WAD
+        // or if
+        //     y * weight + HALF_WAD > type(uint256).max
+        //     <=> weight > 0 and y > (type(uint256).max - HALF_WAD) / weight
+        // or if
+        //     x * (WAD - weight) + y * weight + HALF_WAD > type(uint256).max
+        //     <=> (WAD - weight) > 0 and x > (type(uint256).max - HALF_WAD - y * weight) / (WAD - weight)
+        assembly {
+            z := sub(WAD, weight) // Temporary assignment to save gas.
+            if or(
+                gt(weight, WAD),
+                or(
+                    mul(weight, gt(y, div(MAX_UINT256_MINUS_HALF_WAD, weight))),
+                    mul(z, gt(x, div(sub(MAX_UINT256_MINUS_HALF_WAD, mul(y, weight)), z)))
+                )
+            ) {
+                revert(0, 0)
+            }
+            z := div(add(add(mul(x, z), mul(y, weight)), HALF_WAD), WAD)
+        }
```

Comment:
Don't you think that it is worth factorizing all 3 weightedAvg into one ?

Reply:
Going this way, we could factorize mul & div operations too. I don't know what we prefer to optimize:
- gas cost
- bytecode size

In this particular case, I'd be in favor of minimizing the bytecode size by factorizing mul, div & weighted avg
But this is a matter of another PR IMO
I'm opening an issue to discuss this:

Reply:
I don't think that it is changing the bytecode size, the compiler would probably to a great job here. They only think that it changes is the solidity code size.